mitiq/cdr/clifford_training_data.py

```diff
@@ -0,0 +1,514 @@
+    data, empty_circuit = _circuit_to_array(circuit)
+    mask_rz = data[1, :] == 'rz'
+    rz_circ_data = data[:, mask_rz]
+    mask_not_rz = data[1, :] != 'rz'
+    not_rz_circ_data = data[:, mask_not_rz]
+    mask_non_cliff = _is_clifford_angle(rz_circ_data[2, :])
+    mask_non_cliff = ~mask_non_cliff
+    rz_non_cliff = rz_circ_data[:, mask_non_cliff]
+    mask_cliff = _is_clifford_angle(rz_circ_data[2, :])
+    rz_cliff = rz_circ_data[:, mask_cliff]
+    total_non_cliff = len(rz_non_cliff[0])
+
+    # find all the non-Clifford gates:
+    all_cliff = np.column_stack((not_rz_circ_data, rz_cliff))
+    angles_original_list = []
+    angles_replaced_list = []
+    for n in range(num_training_circuits):
+        # Convert data arry into cirq circuit and append it to the storage
+        #  array:
+        if additional_options:
+            (projected_circuit, angles_original,
+                angles_replaced) = _map_to_near_clifford(
+                rz_non_cliff, all_cliff,
+                empty_circuit,
+                total_non_cliff,
+                fraction_non_clifford,
+                method_select,
+                method_replace,
+                additional_options=additional_options.get(
+                    'additional_options'))
+        else:
+            (projected_circuit, angles_original,
+                angles_replaced) = _map_to_near_clifford(rz_non_cliff,
+                                                         all_cliff,
+                                                         empty_circuit,
+                                                         total_non_cliff,
+                                                         fraction_non_clifford,
+                                                         method_select,
+                                                         method_replace)
+
+        circuits_list.append(projected_circuit)
+        # this information is to make sure the probabilistic methods are
+        #  working as expected:
+        angles_original_list.append(angles_original)
+        angles_replaced_list.append(angles_replaced)
+    return circuits_list, angles_original_list, angles_replaced_list
+
+
+def _map_to_near_clifford(
+    rz_non_cliff: np.ndarray,
+    all_cliff: np.ndarray,
+    empty_circuit: Circuit,
+    total_non_cliff: int,
+    fraction_non_clifford: float,
+    method_select: str = 'random',
+    method_replace: str = 'closest',
+    **additional_options: dict
+) -> (Circuit, List[float], List[float]):
+    """ Function to take the information in some circuit of interest and
+        return a near-Clifford circuit as constructed according to
+        some user defined methods.
+    Args:
+        rz_non_cliff: array containing non-Clifford gates from the circuit of
+                      interest.
+        all_cliff: array containing Clifford gates from the circuit of
+                   interest.
+        empty_circuit: empty circuit strcuture (stores qubit geometry).
```

Comment:
Can you explain more why this is necessary? Qubit geometry isn't really tied to a circuit.